### PR TITLE
lci_strategic_plugin hangs when trying to find entry on green signal

### PR DESCRIPTION
**Pull request by MishkaMN**
# PR Details
## Description

Currently the lci_strategic_plugin is not accounting for an edge case where if there is a green signal in the received signal states, but that signal is behind the vehicle's earliest entry time (EET) possible due to the vehicle's kinematics. 
It is only handling the case if there is no green signal at all. Therefore it may erroneously continue to look for the green signal forever. 

Applied fix is to return the start of TBD state which is the end of the available states with RED signal. Since function can return a timestamp and state that is not GREEN, renamed the function.

Supporting PR to be merged first: https://github.com/usdot-fhwa-stol/autoware.ai/pull/261

## Related GitHub Issue
https://github.com/usdot-fhwa-stol/carma-platform/issues/2227

## Related Jira Key

CDAR-619
https://usdot-carma.atlassian.net/browse/CDAR-619
## How Has This Been Tested?
Simulation PC 1

## Types of changes

- [X] Defect fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

- [X] I have added any new packages to the sonar-scanner.properties file
- [X] My change requires a change to the documentation.
- [X] I have updated the documentation accordingly.
- [X] I have read the [**CONTRIBUTING**](https://github.com/usdot-fhwa-stol/carma-platform/blob/develop/Contributing.md) document.
- [X] I have added tests to cover my changes.
- [X] All new and existing tests passed.

### CODE REVIEW

**Review by adamlm — DISMISSED** [review]

File: lci_strategic_plugin/include/lci_strategic_plugin/lci_strategic_plugin.hpp [hunk 1]
```diff
@@ -575,35 +575,36 @@ class LCIStrategicPlugin : public carma_guidance_plugins::StrategicPlugin
...
    * \param signal CARMATrafficSignal that is responsible for the intersection along route
    * \param minimum_required_green_time Minimum seconds required by other vehicles in front to pass through the intersection first
    *
-   * \return Nearest ROS entry time during green phase.
+   * \return Nearest ROS entry time during green phase, and whether if tbd or not
    * NOTE: TSMO UC2: Algorithm 2. Entering time estimation algorithm for EVs in cooperation Class A when the SPaT plan is fixed-time (defined for C-ADS-equipped vehicles)
    */
 
-  rclcpp::Time get_nearest_green_entry_time(const rclcpp::Time& current_time, const rclcpp::Time& earliest_entry_time, lanelet::CarmaTrafficSignalPtr signal, double minimum_required_green_time = 0.0) const;
+  std::tuple<rclcpp::Time, bool> get_nearest_valid_entry_time(const rclcpp::Time& current_time, const rclcpp::Time& earliest_entry_time, lanelet::CarmaTrafficSignalPtr signal, double minimum_required_green_time = 0.0) const;
```
Comment (adamlm) [inline lci_strategic_plugin/include/lci_strategic_plugin/lci_strategic_plugin.hpp:607]: Consider changing this to `std::optional<rclcpp::Time>` to indicate that a nearest valid entry time does not exist under the given function arguments.

File: lci_strategic_plugin/src/lci_strategic_plugin_algo.cpp [hunk 2]
```diff
@@ -49,7 +49,21 @@ double LCIStrategicPlugin::get_distance_to_accel_or_decel_once (double current_s
   }
 }
 
-rclcpp::Time LCIStrategicPlugin::get_nearest_green_entry_time(const rclcpp::Time& current_time, const rclcpp::Time& earliest_entry_time, lanelet::CarmaTrafficSignalPtr signal, double minimum_required_green_time) const
+rclcpp::Time get_eet_or_tbd(const rclcpp::Time& earliest_entry_time, const lanelet::CarmaTrafficSignalPtr& signal)
```
Comment (adamlm) [inline lci_strategic_plugin/src/lci_strategic_plugin_algo.cpp:52]: See comment about `std::optional`.

File: lci_strategic_plugin/src/lci_strategic_plugin_algo.cpp [hunk 3]
```diff
@@ -241,31 +260,31 @@ std::tuple<rclcpp::Time, bool, bool> LCIStrategicPlugin::get_final_entry_time_an
 
   if (config_.enable_carma_streets_connection ==false || scheduled_enter_time_ == 0) //UC2
   {
-    nearest_green_entry_time = get_nearest_green_entry_time(current_state.stamp, earliest_entry_time, traffic_light) 
-                                          + rclcpp::Duration(EPSILON * 1e9); //0.01sec more buffer since green_light algorithm's timestamp picks the previous signal - Vehicle Estimation
-    is_entry_time_within_green_or_tbd = true; 
+    std::tie(nearest_green_entry_time, in_tbd) = get_nearest_valid_entry_time(current_state.stamp, earliest_entry_time, traffic_light);
```
Comment (adamlm) [inline lci_strategic_plugin/src/lci_strategic_plugin_algo.cpp:263]: See comment about `std::optional`.

**Review by adamlm — CHANGES_REQUESTED** [review]

**Review by adamlm — APPROVED** [review]

**Comment by MishkaMN** [comment]
This was tested locally to be able to build. The autoware.ai's carma-develop branch is pushing to wrong hub.docker.com image tag (carma-develop instead of develop) which is why the CI is failing. Therefore force merging this and force pushing to `usdotfhwastoldev/autoware.ai:develop` with the updated image until the automated process is fixed

**Pull request merged**